### DISCUSSION TITLE
JPA 연관관계 List vs Set

### DISCUSSION THREAD

**Discussion by sangwonsheep** [Kernel360/E2E2-CATCHLINE#42] [General]
```
@OneToMany(mappedBy` = "restaurant")
    private List<ReviewEntity> reviews = new ArrayList<>();

@OneToMany(mappedBy` = "restaurant")
    private Set<ReviewEntity> reviews = new HashSet<>();
```
아래와 같은 코드가 있을 때 Set, List 중에 어떤 방식이 더 좋을까요?

**Comment by sangwonsheep** [comment]
저는 List를 사용하는 방식을 선호합니다.
Set은 기존 데이터를 비교해서 중복이 있는지 확인합니다. 
일반적인 경우는 문제가 없지만, 지연 로딩으로 컬렉션 조회시 문제가 생깁니다.

컬렉션이 아직 초기화 되지 않은 상태에서 값을 넣게 되면 프록시가 강제로 초기화되는 문제가 발생합니다.
그 이유는 중복 데이터가 있는지 비교해야 하는데, 비교를 하기 위해서는 모든 데이터를 로딩해야 하기 때문입니다.(fetch = Eager)

반면에 List는 이런 중복 체크가 필요없기 때문에 추가할 때 프록시 초기화가 발생하지 않습니다.

[인프런 영한님 답변 참고](https://www.inflearn.com/community/questions/321256/collection-type%EC%9C%BC%EB%A1%9C-set-%EB%8C%80%EC%8B%A0-list%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0%EA%B0%80-%EC%9E%88%EB%8A%94%EC%A7%80%EC%9A%94)

**Comment by soeunnPark** [comment]
저도 List를 사용하는 방식이 더 좋다고 생각합니다.

Set의 경우 조인 테이블에 중복되는 레코드가 많아지는걸 방지하기 위해 사용하는 것 같습니다.
다만 상원님이 말씀주신 것처럼 Set 사용시 지연 로딩이 불가하다는 큰 단점이 있기 때문에 오히려 성능이 더 떨어진다는 점에 동의합니다!
